Transformations: Move common SCC utility routines to `utilities`

mlange05 opened the pull request:
~Another small housekeeping change. This one adds a new sub-package `transformations.single_column.util` sub-package for common utilities that may be used across the several SCC sub-stages.~

A small housekeeping change that moves utility methods from `SCCBaseTransformation` to the the common `transformations.utilities` subpackage. This primarily removes the awkward dependency on the `SCCBaseTransformation` class in other SCC sub-components, which provided this previously. 

This change is largely cosmetic (simply copying methods over), except for the `get_local_arrays` utility, which is now used by the demotion and the annotations sub-component. This is in preparation for a more complex refactoring of the annotation sub-component to support multiple vectorisation and annotation schemes in a future PR.

reuterbal reviewed (changes requested):
Thanks, this looks great and removes the indeed awkward cross-referencing to utilities in the Base class, even from outside the SCC transformations.

We should consider creating dedicated tests for these utilities at some point - not mandating this for the PR but maybe create an issue to track this?

Finally, I have paid a little more attention to the new get_local_arrays utility and noticed that we might have a case-sensitivity problem there. Would be good to resolve that (or prove me wrong) before merging.

mlange05 commented on loki/transformations/utilities.py:
```diff
@@ -520,3 +523,167 @@ def apply_to_init_arg(name, arg, expr, mapper):
+        Flag whether to return unique instances of each symbol;
+        default: ``False``
+    """
+    arg_names = tuple(a.name for a in routine.arguments)
```
Yes

mlange05 commented on loki/transformations/utilities.py:
```diff
@@ -520,3 +523,167 @@ def apply_to_init_arg(name, arg, expr, mapper):
+
+    # Filter all variables by argument name to get local arrays
+    arrays = [v for v in variables if isinstance(v, sym.Array)]
+    arrays = [v for v in arrays if v.name not in arg_names]
```
Well spotted! I pushed a fix and tests for this now.

mlange05 commented:
@reuterbal You are, of course, right about the testing. I've added at least rudimentary test coverage for each utility routine. As part of that I spotted an annoying inconsistency, that I also fixed: `get_loop_bounds` would simply return strings and leave the symbol resolution to the upstream call. I've now moved this, which makes testing nicer and removes redundancy. So, please take another look.

reuterbal reviewed (commented):
Thanks, the changes look fantastic and also very much agreed on the change to the return type of get_loop_bounds! Much appreciate the additional testing!

Unless CI throws up some issues this is good to go!
